File: Algorithms/Flow/Algorithms/Part_1_Simple_Algorithms/1.Strange_Algorithm/StrangeAlgorithm.swift
//
//  StrangeAlgorithm.swift
//  Algorithms
//
//  Created by Maksim Ponomarev on 12/5/24.
//

import Foundation

func strangeAlgorithm1(number: Int) {
	var number = number
	while true {
		print(number)
		if number == 1 { break
		}
		if number % 2 == 0 {
			number /= 2
		} else {
			number = number * 3 + 1
		}
	}
}

func strangeAlgorithm2(number: Int) {
	print(number)
	switch number {
	case 1:
		break
		
	default:
		/*
		 let newNumber: Int
		 if number % 2 == 0 {
			newNumber = number / 2
		 } else {
			newNumber = number * 3 + 1
		 }
		 
		 Or
		 */
		
		let newNumber: Int = number % 2 == 0
		? number / 2
		: number * 3 + 1
		
		// Call self with new number
		strangeAlgorithm2(number: newNumber)
	}
	
}

File: Algorithms/Flow/Algorithms/Part_1_Simple_Algorithms/1.Strange_Algorithm/Strange_Algorithm.cpp
//
//  Strange_Algorithm_documentation.cpp
//  Algorithms
//
//  Created by Maksim Ponomarev on 12/5/24.
//

#include <iostream>

using namespace std;

int main() {
	// beter to use long long to not overflow int
	long long n;
	cin >> n;
	while (true) {
		cout << n << endl;
		if (n == 1) break;
		if (n%2 == 0) n/=2;
		else n = n*3+1;
	}
	cout << endl;
}
